// @headless-tree/core/lib/cjs/features/checkboxes/feature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkboxesFeature = void 0;
const utils_1 = require("../../utils");
const types_1 = require("./types");
const errors_1 = require("../../utilities/errors");
const getAllLoadedDescendants = (tree, itemId) => {
    if (!tree.getConfig().isItemFolder(tree.buildItemInstance(itemId))) {
        return [itemId];
    }
    return tree
        .retrieveChildrenIds(itemId)
        .map((child) => getAllLoadedDescendants(tree, child))
        .flat();
};
exports.checkboxesFeature = {
    key: "checkboxes",
    overwrites: ["selection"],
    getInitialState: (initialState) => (Object.assign({ checkedItems: [] }, initialState)),
    getDefaultConfig: (defaultConfig, tree) => {
        var _a;
        const hasAsyncLoader = (_a = defaultConfig.features) === null || _a === void 0 ? void 0 : _a.some((f) => f.key === "async-data-loader");
        if (hasAsyncLoader && !defaultConfig.canCheckFolders) {
            (0, errors_1.throwError)(`!canCheckFolders not supported with async trees`);
        }
        return Object.assign({ setCheckedItems: (0, utils_1.makeStateUpdater)("checkedItems", tree), canCheckFolders: hasAsyncLoader !== null && hasAsyncLoader !== void 0 ? hasAsyncLoader : false }, defaultConfig);
    },
    stateHandlerNames: {
        checkedItems: "setCheckedItems",
    },
    treeInstance: {
        setCheckedItems: ({ tree }, checkedItems) => {
            tree.applySubStateUpdate("checkedItems", checkedItems);
        },
    },
    itemInstance: {
        getCheckboxProps: ({ item }) => {
            const checkedState = item.getCheckedState();
            return {
                onChange: item.toggleCheckedState,
                checked: checkedState === types_1.CheckedState.Checked,
                ref: (r) => {
                    if (r) {
                        r.indeterminate = checkedState === types_1.CheckedState.Indeterminate;
                    }
                },
            };
        },
        toggleCheckedState: ({ item }) => {
            if (item.getCheckedState() === types_1.CheckedState.Checked) {
                item.setUnchecked();
            }
            else {
                item.setChecked();
            }
        },
        getCheckedState: ({ item, tree, itemId }) => {
            const { checkedItems } = tree.getState();
            if (checkedItems.includes(itemId)) {
                return types_1.CheckedState.Checked;
            }
            if (item.isFolder() && !tree.getConfig().canCheckFolders) {
                const descendants = getAllLoadedDescendants(tree, itemId);
                if (descendants.every((d) => checkedItems.includes(d))) {
                    return types_1.CheckedState.Checked;
                }
                if (descendants.some((d) => checkedItems.includes(d))) {
                    return types_1.CheckedState.Indeterminate;
                }
            }
            return types_1.CheckedState.Unchecked;
        },
        setChecked: ({ item, tree, itemId }) => {
            if (!item.isFolder() || tree.getConfig().canCheckFolders) {
                tree.applySubStateUpdate("checkedItems", (items) => [...items, itemId]);
            }
            else {
                tree.applySubStateUpdate("checkedItems", (items) => [
                    ...items,
                    ...getAllLoadedDescendants(tree, itemId),
                ]);
            }
        },
        setUnchecked: ({ item, tree, itemId }) => {
            if (!item.isFolder() || tree.getConfig().canCheckFolders) {
                tree.applySubStateUpdate("checkedItems", (items) => items.filter((id) => id !== itemId));
            }
            else {
                const descendants = getAllLoadedDescendants(tree, itemId);
                tree.applySubStateUpdate("checkedItems", (items) => items.filter((id) => !descendants.includes(id)));
            }
        },
    },
};
