// @headless-tree/core/lib/cjs/test-utils/test-tree.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestTree = void 0;
/* eslint-disable import/no-extraneous-dependencies */
const vitest_1 = require("vitest");
const create_tree_1 = require("../core/create-tree");
const test_tree_do_1 = require("./test-tree-do");
const test_tree_expect_1 = require("./test-tree-expect");
const feature_1 = require("../features/sync-data-loader/feature");
const feature_2 = require("../features/async-data-loader/feature");
const build_proxified_instance_1 = require("../core/build-proxified-instance");
vitest_1.vi.useFakeTimers({ shouldAdvanceTime: true });
class TestTree {
    forSuits(runSuite) {
        vitest_1.describe.for([
            this.suits.sync(),
            this.suits.async(),
            this.suits.proxifiedSync(),
            this.suits.proxifiedAsync(),
        ])("$title", ({ tree }) => {
            tree.resetBeforeEach();
            runSuite(tree);
        });
    }
    get instance() {
        if (!this.treeInstance) {
            this.treeInstance = (0, create_tree_1.createTree)(this.config);
            this.treeInstance.rebuildTree();
        }
        return this.treeInstance;
    }
    constructor(config) {
        this.config = config;
        this.do = new test_tree_do_1.TestTreeDo(this);
        this.expect = new test_tree_expect_1.TestTreeExpect(this);
        this.treeInstance = null;
        this.asyncDataLoaderImp = {
            getItem: (id) => __awaiter(this, void 0, void 0, function* () {
                yield new Promise((r) => {
                    r.debugName = `Loading getItem ${id}`;
                    TestTree.asyncLoaderResolvers.push(r);
                });
                return id;
            }),
            getChildren: (id) => __awaiter(this, void 0, void 0, function* () {
                yield new Promise((r) => {
                    r.debugName = `Loading getChildren ${id}`;
                    TestTree.asyncLoaderResolvers.push(r);
                });
                return [`${id}1`, `${id}2`, `${id}3`, `${id}4`];
            }),
        };
        this.suits = {
            sync: () => ({
                tree: this.withFeatures(feature_1.syncDataLoaderFeature),
                title: "Synchronous Data Loader",
            }),
            async: () => ({
                tree: this.withFeatures(feature_2.asyncDataLoaderFeature).with({
                    dataLoader: this.asyncDataLoaderImp,
                }),
                title: "Asynchronous Data Loader",
            }),
            proxifiedSync: () => ({
                tree: this.withFeatures(feature_1.syncDataLoaderFeature).with({
                    instanceBuilder: build_proxified_instance_1.buildProxiedInstance,
                }),
                title: "Proxified Synchronous Data Loader",
            }),
            proxifiedAsync: () => ({
                tree: this.withFeatures(feature_2.asyncDataLoaderFeature).with({
                    instanceBuilder: build_proxified_instance_1.buildProxiedInstance,
                }),
                title: "Proxified Asynchronous Data Loader",
            }),
        };
    }
    static resolveAsyncLoaders() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            do {
                (_a = TestTree.asyncLoaderResolvers.shift()) === null || _a === void 0 ? void 0 : _a();
                yield new Promise((r) => {
                    setTimeout(r);
                });
            } while (TestTree.asyncLoaderResolvers.length);
        });
    }
    resolveAsyncVisibleItems() {
        return __awaiter(this, void 0, void 0, function* () {
            this.instance.getItems();
            yield TestTree.resolveAsyncLoaders();
            this.instance.getItems().forEach((i) => i.getItemName());
            yield TestTree.resolveAsyncLoaders();
        });
    }
    static default(config) {
        return new TestTree(Object.assign({ rootItemId: "x", createLoadingItemData: () => "loading", dataLoader: {
                getItem: (id) => id,
                getChildren: (id) => [`${id}1`, `${id}2`, `${id}3`, `${id}4`],
            }, getItemName: (item) => item.getItemData(), indent: 20, isItemFolder: (item) => item.getItemMeta().level < 2, initialState: {
                expandedItems: ["x1", "x11"],
            }, features: [] }, config));
    }
    with(config) {
        return new TestTree(Object.assign(Object.assign({}, this.config), config));
    }
    resetBeforeEach() {
        (0, vitest_1.beforeEach)(() => __awaiter(this, void 0, void 0, function* () {
            yield this.createTestCaseTree();
        }));
    }
    createTestCaseTree() {
        return __awaiter(this, void 0, void 0, function* () {
            this.reset();
            vitest_1.vi.clearAllMocks();
            // trigger instance creation
            // eslint-disable-next-line @typescript-eslint/no-unused-expressions
            this.instance;
            yield this.resolveAsyncVisibleItems();
            return this;
        });
    }
    withFeatures(...features) {
        var _a;
        return this.with({
            features: [...((_a = this.config.features) !== null && _a !== void 0 ? _a : []), ...features],
        });
    }
    mockedHandler(handlerName) {
        var _a;
        const mock = vitest_1.vi.fn();
        if (this.treeInstance) {
            (_a = this.treeInstance) === null || _a === void 0 ? void 0 : _a.setConfig((prev) => (Object.assign(Object.assign({}, prev), { [handlerName]: mock })));
        }
        else {
            this.config[handlerName] = mock;
        }
        return mock;
    }
    item(itemId) {
        return this.instance.getItemInstance(itemId);
    }
    reset() {
        this.treeInstance = null;
        TestTree.asyncLoaderResolvers = [];
    }
    debug() {
        console.log(this.instance
            .getItems()
            .map((item) => [
            "  ".repeat(item.getItemMeta().level),
            '"',
            item.getItemName(),
            '"',
        ].join(""))
            .join("\n"));
    }
    setElementBoundingBox(itemId, bb = {
        left: 0,
        width: 100,
        top: 0,
        height: 20,
    }) {
        this.instance.registerElement({
            getBoundingClientRect: () => ({
                left: 0,
                width: 100,
                top: 0,
                height: 10000,
            }),
        });
        this.instance.getItemInstance(itemId).registerElement({
            getBoundingClientRect: () => bb,
        });
    }
    static dragEvent(clientX = 1000, clientY = 0) {
        return {
            preventDefault: vitest_1.vi.fn(),
            stopPropagation: vitest_1.vi.fn(),
            dataTransfer: {
                setData: vitest_1.vi.fn(),
                getData: vitest_1.vi.fn(),
                dropEffect: "unchaged-from-test",
            },
            clientX,
            clientY,
        };
    }
    createTopDragEvent(indent = 0) {
        return TestTree.dragEvent(indent * 20, 1);
    }
    createBottomDragEvent(indent = 0) {
        return TestTree.dragEvent(indent * 20, 19);
    }
}
exports.TestTree = TestTree;
TestTree.asyncLoaderResolvers = [];
